src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -44,6 +205,469 @@ private void BreakPointWithDebuggerFuncEval()
+                        _expectedType = objType;
+                    }
+                    return _fetchForExpectedType.Fetch(obj);
+                }
+
+                /// <summary>
+                /// A public field that can be used to form a linked list.   
+                /// </summary>
+                public PropertySpec Next;
+
+                #region private
+                /// <summary>
+                /// PropertyFetch is a helper class.  It takes a PropertyInfo and then knows how
+                /// to efficiently fetch that property from a .NET object (See Fetch method).  
+                /// It hides some slightly complex generic code.  
+                /// </summary>
+                class PropertyFetch
+                {
+                    /// <summary>
+                    /// Create a property fetcher from a .NET Reflection PropertyInfo class that
+                    /// represents a property of a paritcular type.  
+                    /// </summary>
+                    public static PropertyFetch FetcherForProperty(PropertyInfo propertyInfo)
+                    {
+                        if (propertyInfo == null)
+                            return new PropertyFetch();     // returns null on any fetch.
+
+                        var typedPropertyFetcher = typeof(TypedFetchProperty<,>);
+                        var instantiatedTypedPropertyFetcher = typedPropertyFetcher.GetTypeInfo().MakeGenericType(
+                            propertyInfo.DeclaringType, propertyInfo.PropertyType);
+                        return (PropertyFetch)Activator.CreateInstance(instantiatedTypedPropertyFetcher, propertyInfo);
+                    }
+
+                    /// <summary>
+                    /// Given an object, fetch the property that this propertyFech represents. 
+                    /// </summary>
+                    public virtual object Fetch(object obj) { return null; }
+
+                    #region private 
+
+                    class TypedFetchProperty<Tobject, Tproperty> : PropertyFetch
+                    {
+                        public TypedFetchProperty(PropertyInfo property)
+                        {
+                            _propertyFetch = (Func<Tobject, Tproperty>)property.GetMethod.CreateDelegate(typeof(Func<Tobject, Tproperty>));
+                        }
+                        public override object Fetch(object obj)
+                        {
+                            return _propertyFetch((Tobject)obj);
+                        }
+                        private Func<Tobject, Tproperty> _propertyFetch;
+                    }
+                    #endregion
+                }
+
+                string _propertyName;
+                Type _expectedType;
+                PropertyFetch _fetchForExpectedType;
+                #endregion
+            }
+
+            string _outputName;
+            PropertySpec _fetches;
```

Comment:
Nit: explicit visibility for all of these fields.  I'll stop commenting on these :smile: